keps/sig-storage/3476-volume-group-snapshot/README.md

```diff
@@ -360,10 +373,20 @@ Type VolumeGroupSnapshotStatus struct {
         // +optional
         Error *VolumeSnapshotError
 
-        // List of volume snapshot refs
-        // The max number of snapshots in the group is 100
-        // +optional
-        VolumeSnapshotRefList []core_v1.ObjectReference
+	// VolumeSnapshotRefList is the list of PVC and VolumeSnapshot pairs that
+	// is part of this group snapshot.
+	// The maximum number of allowed snapshots in the group is 100.
+	// +optional
+	PVCVolumeSnapshotRefList []PVCVolumeSnapshotPair
+}
+
+// PVCVolumeSnapshotPair defines a pair of a PVC reference and a Volume Snapshot Reference
+type PVCVolumeSnapshotPair struct {
+	// PersistentVolumeClaimRef is a reference to the PVC this pair is referring to
+	PersistentVolumeClaimRef core_v1.LocalObjectReference
+
+	// VolumeSnapshotRef is a reference to the VolumeSnapshot this pair is referring to
+	VolumeSnapshotRef core_v1.LocalObjectReference
```

Comment:
Why do we need these items in VolumeGroupSnapshotStatus?

* relation VolumeSnapshot -> VolumeGroupSnapshot is already established in volumeSnapshot `status.volumeGroupSnapshotName`. If we repeat it here, it will be only error prone.

* The same for PVC, volumeSnapshot's `spec.source.persistentVolumeClaimName`

Reply:
This change was made to address this issue: https://github.com/kubernetes-csi/external-snapshotter/issues/969
We need to be able to restore from a VolumeGroupSnapshot.

Reply:
I don't understand it.  The issue there is:

> 1. Create their PVCs for their app, and label them
> 2. Create a volumegroupsnapshot with the label selector
> 3. Now when they want to restore, they need to look at the volumegroupsnaphot, then the volumegroupsnapshotcontents, then map the PV list to the volumesnapshotcontents list, and then find the volumesnapshots from the volumesnapshotcontents.  This also assumes they're keeping track of the PVs that were associated with the original PVCs from step 1 (note that for step1 & 2 they are dealing with PVCs only, not the underlying PVs).  At this point when they're trying to restore their data, the original PVCs may also have been deleted.
> 4. Now they can startup their app that uses the PVCs - if at this point the wrong volumesnapshot has been restored to the wrong pvc name this would be very problematic. 

But at the point 3, they can list VolumeSnapshots that are part of a group and see to which PVCs they refer to. We don't need to repeat that information in `VolumeGroupSnapshotStatus`. Am I missing anything?

Reply:
ugh, did the API in [1068](https://github.com/kubernetes-csi/external-snapshotter/pull/1068/files) go through API review? It's a Kubernetes API

Reply:
I did consult @msau42 about this change.

Reply:
A user wants to be able to create PVCs from the VolumeGroupSnapshot.
A user will only be looking at VolumeGroupSnapshot that is in the namespace scope.
User won’t have access to VolumeGroupSnapshotContent.
So we want to provide sufficient information for a user to create PVCs from VolumeSnapshots by inspecting a VolumeGroupSnapshot API object.

Reply:
I see what is wrong. VolumeSnapshots that come out of a VolumeGroupSnapshot do not have filed `spec.source.persistentVolumeClaimName`. IMO that's a bug.

Is there any technical reason it cannot be filed there? I can see the external-snapshotter finds the PV when [updating VolumeSnapshotContent.status](https://github.com/jsafrane/external-snapshotter/blob/209a5b4372166e7b0b209088a194c26ef159d247/pkg/sidecar-controller/groupsnapshot_helper.go#L669-L669) just after [creating VolumeSnapshot](https://github.com/jsafrane/external-snapshotter/blob/209a5b4372166e7b0b209088a194c26ef159d247/pkg/sidecar-controller/groupsnapshot_helper.go#L499-L499). Why can't find it *before* creating VolumeSnapshots?

In the end, we have two sources of VolumeSnapshot PVC: one in the VolumeSnapshot itself and another one if VolumeGroupSnapshot status. IMHO this is inconsistent, error prone and will cause confusion.

Reply:
It is not a bug. It is designed this way. Let me explain.
- User creates a VolumeGroupSnapshot API object.
- A CSI driver creates a volume group snapshot in one gRPC call. As a result, it returns a group snapshot id and a list of snapshots (with source volume id and snapshot id).
- The group controller logic will then create VolumeSnapshotContent that points to the existing snapshot id and create VolumeSnapshot to bound to the pre-provisioned snapshot.
- Since it is a pre-provisioned snapshot, VolumeSnapshot.spec.source will have VolumeSnapshotContentName, not PersistentVolumeClaimName.

Reply:
https://github.com/kubernetes-csi/external-snapshotter/blob/master/client/apis/volumesnapshot/v1/types.go#L106

Reply:
> Since it is a pre-provisioned snapshot, VolumeSnapshot.spec.source will have VolumeSnapshotContentName, not PersistentVolumeClaimName.

That's where we disagree. I think it should have PersistentVolumeClaimName. It *is* a dynamically provisioned snapshot. Just using a slightly different workflow.

Reply:
We discussed this today. We can set PVCName in this case. We still need to have PVCVolumeSnapshotRefList in the Status for restore as we also need to support static provisioning.